src/VecSim/algorithms/hnsw/hnsw.h

```diff
@@ -903,24 +954,60 @@ void HNSWIndex<DataType, DistType>::SwapLastIdWithDeletedId(idType element_inter
 
     // swap the last_id level 0 data, and invalidate the deleted id's data
     memcpy(data_level0_memory_ + element_internal_id * size_data_per_element_ + offsetLevel0_,
-           data_level0_memory_ + max_id * size_data_per_element_ + offsetLevel0_,
+           data_level0_memory_ + cur_element_count * size_data_per_element_ + offsetLevel0_,
            size_data_per_element_);
-    memset(data_level0_memory_ + max_id * size_data_per_element_ + offsetLevel0_, 0,
+    memset(data_level0_memory_ + cur_element_count * size_data_per_element_ + offsetLevel0_, 0,
            size_data_per_element_);
 
     // swap pointer of higher levels links
-    linkLists_[element_internal_id] = linkLists_[max_id];
-    linkLists_[max_id] = nullptr;
+    linkLists_[element_internal_id] = linkLists_[cur_element_count];
+    linkLists_[cur_element_count] = nullptr;
 
     // swap top element level
-    element_levels_[element_internal_id] = element_levels_[max_id];
-    element_levels_[max_id] = HNSW_INVALID_LEVEL;
+    element_levels_[element_internal_id] = element_levels_[cur_element_count];
+    element_levels_[cur_element_count] = HNSW_INVALID_LEVEL;
 
-    if (max_id == this->entrypoint_node_) {
+    if (cur_element_count == this->entrypoint_node_) {
         this->entrypoint_node_ = element_internal_id;
     }
 }
 
+template <typename DataType, typename DistType>
+template <bool with_timeout>
+void HNSWIndex<DataType, DistType>::greedilySearchLevel(const void *vector_data, size_t level,
```

Comment:
I think the name is a bit confusing, since also the function that returns the top candidates is doing greedy search. Consider at least add a very good explenation about the function purpose